### PR TITLE
GH-67: robot dependent widgets

### LINKED ISSUE #67: Settings | Widgets should be robot-dependent
Currently, the user can customize widgets in general and that view state is displayed for all robots. 
However, some robots may not have Video camera and as a result, the Screen widget is unnecessary. Or another example, the user don't want to see 'Robot's speed' widget on the main page for one robot, but it's required for another one.

Thus, it should be possible to configure widgets for each robot separately.

**Note:** it may require updates on UI and in designs by merging Widgets settings with the main robot's settings.

Screenshot 2023-04-10 at 14 12 33

### PR DESCRIPTION
**Pull request by alex-waverley**
updated logic and UI for settings tab:
- every robot has it's own set of widgets
- if user makes changes in settings and without saving them tries to go to another tab - modal window appears and suggests to save changes
Screenshot 2023-04-13 at 12 36 39

Closes #67

### CODE REVIEW

**Review by sskorol — DISMISSED** [review]

File: frontend/src/components/streamingModal/index.js [hunk 1]
```diff
@@ -1,51 +1,31 @@
...
-          <Button sx={{ textTransform: 'uppercase' }} isLight onClick={onStopStreamingConfirm}>yes, I want to end streaming</Button>
-          <Button sx={{ textTransform: 'uppercase', marginLeft: '24px' }} onClick={onStopStreamingCancel}>no, I don’t </Button>
-        </Grid>
-      </Box>
-    </Modal>
+  return currentRobot && <Modal
+    showModal={actionModalName === ActionModalName.STREANMING && showActionModal}
+    onCancel={onCancelActionModal}
+    onConfirm={onConfirmActionModal}
+    confirmButtonText='yes, I want to end streaming'
```
Comment (sskorol) [inline frontend/src/components/streamingModal/index.js:20]: Shouldn't it start with a capital letter?

File: frontend/src/store/ActionModal/index.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  @observable actionModalName = null
+  @observable showActionModal = false
+  @observable actionModalConfirmCallback = null
+  @observable actionModalCancelCallback = null
+
+  @action
+  onShowActionModal  = ({ actionModalName, actionModalConfirmCallback, actionModalCancelCallback }) => {
+    this.actionModalName = actionModalName
+    this.showActionModal = true
+    if(typeof actionModalConfirmCallback === 'function') {
```
Comment (sskorol) [inline frontend/src/store/ActionModal/index.js:22]: seems like a missing formatting within this file

File: frontend/src/store/ActionModal/index.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      this.actionModalConfirmCallback = actionModalConfirmCallback
+    }
+    if(typeof actionModalCancelCallback === 'function') {
```
Comment (sskorol) [inline frontend/src/store/ActionModal/index.js:25]: What's the other type it could be?

File: frontend/src/util.js [hunk 4]
```diff
@@ -13,3 +13,11 @@ export function formatTime(d, timeFormat) {
     ''
   )
 }
+
+export async function delay(cb, time) {
+  return new Promise(function(resolve, _reject) {
```
Comment (sskorol) [inline frontend/src/util.js:18]: why not an arrow function?

**Review by sskorol — CHANGES_REQUESTED** [review]

File: frontend/src/components/streamingModal/index.js (hunk 1 above)
Comment (alex-waverley) [reply frontend/src/components/streamingModal/index.js:20]: button's text is capitalized by css

File: frontend/src/store/ActionModal/index.js (hunk 3 above)
Comment (alex-waverley) [reply frontend/src/store/ActionModal/index.js:25]: `function` is the only acceptable type

File: frontend/src/util.js (hunk 4 above)
Comment (alex-waverley) [reply frontend/src/util.js:18]: done

File: frontend/src/store/ActionModal/index.js (hunk 2 above)
Comment (alex-waverley) [reply frontend/src/store/ActionModal/index.js:22]: done

File: frontend/src/components/dashboard/index.js [hunk 5]
```diff
@@ -18,14 +18,14 @@ import useMediaQuery from '@mui/material/useMediaQuery'
...
     rosStore: { isWSConnected, isTeleopReady },
   } = useStore()
 
   const connected = isWSConnected && isTeleopReady
-  const isBatteryWidgetSelected = widgets.find((widget) => widget.name === 'battery').selected
-  const isRobotSpeedWidgetSelected = widgets.find((widget) => widget.name === 'speed').selected
-  const isAdditionalActionsWidgetSelected = widgets.find((widget) => widget.name === 'actions').selected
+  const isBatteryWidgetSelected = currentRobot?.widgets.find((widget) => widget.name === 'battery')?.selected
+  const isRobotSpeedWidgetSelected = currentRobot?.widgets.find((widget) => widget.name === 'speed')?.selected
+  const isAdditionalActionsWidgetSelected = currentRobot?.widgets.find((widget) => widget.name === 'actions')?.selected
```
[the remark this answers is not in the document]
Comment (alex-waverley) [reply frontend/src/components/dashboard/index.js:28]: done

File: frontend/src/components/robotPropertiesSettings/index.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  const handleSpeedMinChange = (e) => setSpeedMin(e.target.value)
+  const handleSpeedMaxChange = (e) => setSpeedMax(e.target.value)
+
+  const onSubmit = (e) => {
+    e.preventDefault()
+
+    sendRobotSettings()
+  }
+
+  const sendRobotSettings = useCallback(() => {
```
[the remark this answers is not in the document]
Comment (alex-waverley) [reply frontend/src/components/robotPropertiesSettings/index.js:44]: yes, widget states are updated dynamically and the rest settings (robot name, speed) is updated when user clicks `apply` button

**Review by sskorol — APPROVED** [review]

**Pull request merged**